### PR TITLE
sub-features for node level

### PR DESCRIPTION
**Pull request by guolinke**
implement #2315

### CODE REVIEW

File: include/LightGBM/config.h [hunk 1]
```diff
@@ -268,6 +268,12 @@ struct Config {
   // desc = random seed for bagging
   int bagging_seed = 3;
 
+  // alias = sub_feature_bynode, colsample_bytree_bynode
+  // desc = set to ``true`` will randomly select part of features for each node
```
Comment (StrikerRUS) [inline include/LightGBM/config.h]: It seems to me that `set to` -> `set this to` and `will` -> `to` would be better wording.

File: include/LightGBM/config.h [hunk 2]
```diff
@@ -268,6 +268,12 @@ struct Config {
...
+  // desc = set to ``false`` will randomly select part of features for each true (use the same sub features for each tree)
```
Comment (StrikerRUS) [inline include/LightGBM/config.h]: Typo in `each true (use`.

File: include/LightGBM/config.h [hunk 3]
```diff
@@ -268,6 +268,12 @@ struct Config {
...
+  // desc = **Note**: set to ``false`` can speed up the training linearly
```
Comment (StrikerRUS) [inline include/LightGBM/config.h]: I think it's better to invert the meaning of this note, because the default value is `false`. Inattentive user may expect the speedup by setting this param to `false`, but nothing will happen because it has been already `false`.

**Comment by guolinke** [comment]
@BlindApe could you try this PR could provide some feedbacks?

**Comment by BlindApe** [comment]
Thank you @guolinke for this feature, and sorry for delay. I was on holidays and disconnected.
I've tried to use this and seems perhaps I'm doing something badly. The results are the same:

> rm(list = ls(all = TRUE))
> require(lightgbm)
> require(data.table)
> data(agaricus.train, package = "lightgbm")
> train <- agaricus.train
> dtrain <- lgb.Dataset(train$data, label = train$label)
> data(agaricus.test, package = "lightgbm")
> test <- agaricus.test
> dtest <- lgb.Dataset.create.valid(dtrain, test$data, label = test$label)
> params <- list(objective = "regression", metric = "l2")
> valids <- list(test = dtest)
> model <- lgb.train(params, dtrain, num_round = 5,  valids,  min_data = 1,  learning_rate = 0.05, feature_fraction_bynode = FALSE)
[LightGBM] [Info] Total Bins 369
[LightGBM] [Info] Number of data: 6513, number of used features: 116
[LightGBM] [Info] Start training from score 0.482113
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[1]:	test's l2:0.225323 
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[2]:	test's l2:0.203354 
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[3]:	test's l2:0.183527 
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[4]:	test's l2:0.165633 
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[5]:	test's l2:0.149484 
> 
> model <- lgb.train(params, dtrain, num_round = 5, valids, min_data = 1, learning_rate = 0.05, feature_fraction_bynode = TRUE)
[LightGBM] [Info] Total Bins 369
[LightGBM] [Info] Number of data: 6513, number of used features: 116
[LightGBM] [Info] Start training from score 0.482113
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[1]:	test's l2:0.225323 
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[2]:	test's l2:0.203354 
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[3]:	test's l2:0.183527 
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[4]:	test's l2:0.165633 
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[5]:	test's l2:0.149484

**Comment by guolinke** [comment]
@BlindApe it seems you forgot to set `feature_fraction` .

**Comment by BlindApe** [comment]
When I change the parameter to a fraction (0.1):
> model <- lgb.train(params,  dtrain, 5, valids, min_data = 1,  learning_rate = 0.05, feature_fraction_bynode = 0.1, early_stopping_rounds = 10)
[LightGBM] [Fatal] Parameter feature_fraction_bynode should be "true"/"+" or "false"/"-", got "0.1"
[LightGBM] [Fatal] Parameter feature_fraction_bynode should be "true"/"+" or "false"/"-", got "0.1"
[LightGBM] [Fatal] Parameter feature_fraction_bynode should be "true"/"+" or "false"/"-", got "0.1"
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[1]:	test's l2:0.134909 
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[2]:	test's l2:0.121755 
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[3]:	test's l2:0.109884 
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[4]:	test's l2:0.0991705 
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[5]:	test's l2:0.0895013 

And for differente fraction:

> model <- lgb.train(params, dtrain, 5, valids, min_data = 1, learning_rate = 0.05, feature_fraction_bynode = 0.2)
[LightGBM] [Fatal] Parameter feature_fraction_bynode should be "true"/"+" or "false"/"-", got "0.2"
[LightGBM] [Fatal] Parameter feature_fraction_bynode should be "true"/"+" or "false"/"-", got "0.2"
[LightGBM] [Fatal] Parameter feature_fraction_bynode should be "true"/"+" or "false"/"-", got "0.2"
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[1]:	test's l2:0.080775 
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[2]:	test's l2:0.0728994 
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[3]:	test's l2:0.0657917 
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[4]:	test's l2:0.059377 
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[5]:	test's l2:0.0535878 

In this case results are differents.
So seems I haven't clear if the parameter need to be boolean or a fraction.
The best option is a fraction and then set the number of features to use in each node to number of features in the tree (determined by feature_fraction) * feature_fraction_bynode and rounding to upper integer. In this way using a very small feature_fraction_bynode we can get that only a feature is sampled by node and simulate an extra-tree in rf booster.

**Comment by BlindApe** [comment]
@guolinke If I set feature_fraction to 1 I get the same results for feature_fraction_bynode TRUE or FALSE

**Comment by guolinke** [comment]
@BlindApe yeah, feature_fraction_bynode is a flag to open subcol by node. 
feature_fraction is the used ratio of features. 

I add the test case, to test the different feature fraction when enabling subcol_by_node.

**Comment by guolinke** [comment]
@BlindApe the test passed. So it should be different when with different feature_fraction.

File: tests/python_package_test/test_engine.py [hunk 4]
```diff
@@ -1590,3 +1590,33 @@ def constant_metric(preds, train_data):
...
+                        evals_result=evals_result)
+        ret = log_loss(y_test, gbm.predict(X_test))
+        self.assertLess(ret, 0.13)
+        self.assertAlmostEqual(evals_result['valid_0']['binary_logloss'][-1], ret, places=5)
+        params['feature_fraction'] = 0.5
+        gbm2 = lgb.train(params, lgb_train,
+                         num_boost_round=25,
+                         valid_sets=lgb_eval,
+                         verbose_eval=False,
+                         evals_result=evals_result)
```
Comment (StrikerRUS) [inline tests/python_package_test/test_engine.py]: Seems that valid set is not used in the further code.

**Comment by BlindApe** [comment]
OK, I undestand now.
I've tested the functionality and works.
Thank you!

rm(list = ls(all = TRUE))
require(lightgbm)
require(data.table)
data(agaricus.train, package = "lightgbm")
train <- agaricus.train
dtrain <- lgb.Dataset(train$data, label = train$label)
data(agaricus.test, package = "lightgbm")
test <- agaricus.test
dtest <- lgb.Dataset(test$data, label = test$label)
valids <- list(train = dtrain, test = dtest)
params <- list(objective = "binary", feature_fraction = 0.3, learning_rate = 0.05, feature_fraction_bynode = FALSE)
gbm1 <- lgb.train(params = params, data = dtrain, num_round = 5, valids = valids)
[LightGBM] [Warning] Starting from the 2.1.2 version, default value for the "boost_from_average" parameter in "binary" objective is true.
This may cause significantly different results comparing to the previous versions of LightGBM.
Try to set boost_from_average=false, if your old models produce bad results
[LightGBM] [Info] Number of positive: 3140, number of negative: 3373
[LightGBM] [Info] Total Bins 342
[LightGBM] [Info] Number of data: 6513, number of used features: 107
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.482113 -> initscore=-0.071580
[LightGBM] [Info] Start training from score -0.071580
[1]:	train's binary_logloss:0.646381	test's binary_logloss:0.646845 
[2]:	train's binary_logloss:0.602521	test's binary_logloss:0.603166 
[3]:	train's binary_logloss:0.563393	test's binary_logloss:0.564101 
[4]:	train's binary_logloss:0.526795	test's binary_logloss:0.527475 
[5]:	train's binary_logloss:0.493253	test's binary_logloss:0.493936 

params <- list(objective = "binary", feature_fraction = 0.3, learning_rate = 0.05, feature_fraction_bynode = TRUE)
gbm2 <- lgb.train(params = params, data = dtrain, num_round = 5, valids = valids)
[LightGBM] [Warning] Starting from the 2.1.2 version, default value for the "boost_from_average" parameter in "binary" objective is true.
This may cause significantly different results comparing to the previous versions of LightGBM.
Try to set boost_from_average=false, if your old models produce bad results
[LightGBM] [Info] Number of positive: 3140, number of negative: 3373
[LightGBM] [Info] Total Bins 342
[LightGBM] [Info] Number of data: 6513, number of used features: 107
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.482113 -> initscore=-0.071580
[LightGBM] [Info] Start training from score -0.071580
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[1]:	train's binary_logloss:0.64433	test's binary_logloss:0.644204 
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[2]:	train's binary_logloss:0.600371	test's binary_logloss:0.600178 
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[3]:	train's binary_logloss:0.560376	test's binary_logloss:0.560181 
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[4]:	train's binary_logloss:0.523975	test's binary_logloss:0.523874 
[5]:	train's binary_logloss:0.490471	test's binary_logloss:0.490451
